### PR TITLE
[IMP] portal: redesign

### PR DESCRIPTION
**Pull request by chgo-odoo**
This PR introduces a general redesign of the Portal user interface in order
to give it that fresh and sleek vibe we also introduced in the backend. 

A list of the different changes can be found below. It includes redesign
of some elements as much as layout modification, or element additions.

---

# Main key changes

### Introducing pictograms

It is now a common practice to use pictograms across the UI when it comes
to illustrate a concept related to a module but not the module itself.
With the new cards layout, we implemented these pictograms to illustrate
what each card is about and provides the user some context.

image

### Layout and design update

With this PR, portal is receiving several design changes, as the card layout
on the home page as mentioned earlier. We also changed the general UI,
removing  custom `background-color` and aiming to something clean and sleek
in the same vibe as what we introduced in the backend.

### Improved responsiveness

With this new fluid layout, Portal gets a better responsiveness and benefits
from a better layout at all media query, with different structures, always
showcasing the essential information.

### SCSS files revamp

We also took that opportunity to clean our SCSS files and remove custom
declarations that were not needed anymore. We tend to reduce the amount
of custom CSS and making use of our BS or Odoo classes.

task-3046004

Co-authored-by: fdamhaut (FLDA) <flda@odoo.com>
Co-authored-by: stefanorigano (SRI) <sri@odoo.com>

### CODE REVIEW

**Review by stefanorigano — COMMENTED** [review]
Hello @chgo-odoo , thanks 👍 
Some INFO comments for you to keep track about what I'm doing in the branch (you don't have to apply my suggestion, I'll push some fixup shortly)

**Comment by chgo-odoo** [comment]
@stefanorigano Could you have a look please ?

File: addons/portal/static/src/scss/portal.scss [hunk 1]
```diff
@@ -13,10 +13,10 @@ $o-portal-mobile-toolbar-border: $border-color;
 $o-portal-mobile-toolbar-bg: $gray-200;
 
 // Portal Tables
-$o-portal-table-th-pt: map-get($spacers, 2) !default; // bts4 pt-2
-$o-portal-table-th-pb: map-get($spacers, 2) !default; // bts4 pb-2
-$o-portal-table-td-pt: map-get($spacers, 1) !default; // bts4 pt-1
-$o-portal-table-td-pb: map-get($spacers, 1) !default; // bts4 pb-1
+$o-portal-table-th-pt: map-get($spacers, 1) !default;
+$o-portal-table-th-pb: map-get($spacers, 1) !default;
+$o-portal-table-td-pt: map-get($spacers, 2) !default; // bts4 pt-2
```
Comment (stefanorigano) [inline addons/portal/static/src/scss/portal.scss:18]: You can remove these comments 
```suggestion
$o-portal-table-td-pt: map-get($spacers, 2) !default;
```

File: addons/portal/static/src/scss/portal.scss [hunk 2]
```diff
@@ -31,6 +31,10 @@ body {
     direction: ltr;
 }
 
+th{
```
Comment (stefanorigano) [inline addons/portal/static/src/scss/portal.scss:34]: Let's remove these lines ;)
Tables design should be defined at BS level. Also this setting may mismatch user's preferences.

Anyway, if we want to customize Portal's tables only, next time we should target `.o_portal_my_doc_table th` specifically  ;)

File: addons/portal/static/src/scss/portal.scss [hunk 3]
```diff
@@ -13,10 +13,10 @@ $o-portal-mobile-toolbar-border: $border-color;
...
-$o-portal-table-th-pb: map-get($spacers, 2) !default; // bts4 pb-2
-$o-portal-table-td-pt: map-get($spacers, 1) !default; // bts4 pt-1
-$o-portal-table-td-pb: map-get($spacers, 1) !default; // bts4 pb-1
+$o-portal-table-th-pt: map-get($spacers, 1) !default;
+$o-portal-table-th-pb: map-get($spacers, 1) !default;
+$o-portal-table-td-pt: map-get($spacers, 2) !default; // bts4 pt-2
+$o-portal-table-td-pb: map-get($spacers, 2) !default; // bts4 pb-2
 
 // Portal custom bg color
 $o-portal-bg-color: desaturate($gray-200, 100%);
```
Comment (stefanorigano) [inline addons/portal/static/src/scss/portal.scss:22]: This variable (and the following `$o-portal-use-default-colors` one) is probably not needed anymore, could you check? Eventually these lines should be removed.
Don't forget to check if the variables are used by other modules. 

Removing these variables should be in a separate commit with a proper description ;)

File: addons/portal/static/src/scss/portal.scss [hunk 4]
```diff
@@ -405,6 +360,7 @@ img, .media_iframe_video, .o_image {
     }
 }
 
+
```
Comment (stefanorigano) [inline addons/portal/static/src/scss/portal.scss:363]: Unnecessary change

File: addons/portal/views/portal_templates.xml [hunk 5]
```diff
@@ -29,8 +29,8 @@
             </t>
         </xpath>
         <xpath expr="//div[@id='wrapwrap']/main/t[@t-out='0']" position="before">
-            <div t-if="o_portal_fullwidth_alert" class="alert alert-info alert-dismissible rounded-0 fade show d-print-none css_editable_mode_hidden">
-                <div class="container">
+            <div  t-if="o_portal_fullwidth_alert" class="container mt-3">
```
Comment (stefanorigano) [inline addons/portal/views/portal_templates.xml:32]: Typo ;)

```suggestion
            <div t-if="o_portal_fullwidth_alert" class="container mt-3">
```

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Almost there, courage 💪

**Review by stefanorigano — APPROVED** [review]
LGTM 👍

File: addons/account_payment/views/account_portal_templates.xml [hunk 6]
```diff
@@ -60,6 +60,14 @@
     </template>
 
     <template id="portal_invoice_page_inherit_payment" name="Payment on My Invoices" inherit_id="account.portal_invoice_page">
+        <xpath expr="//t[@t-call='portal.portal_record_sidebar']//h2[1]" position="after">
+            <div t-if="invoice.payment_state in ('paid', 'in_payment')" class="position-sm-absolute position-lg-relative me-auto me-lg-0 top-0 end-0 rounded text-bg-success fs-6 fw-normal badge">
+                <i class="fa fa-fw fa-check-circle"/> Paid
+            </div>
+            <div t-else="pending_txs" class="position-sm-absolute position-lg-relative me-auto me-lg-0 top-0 end-0 rounded text-bg-info fw-normal fs-6 badge" >
```
Comment (AntoineVDV) [inline addons/account_payment/views/account_portal_templates.xml:67]: I assume you intended to use `t-elif` rather than `t-else`, as `t-else` doesn't care about the provided value.

Comment (AntoineVDV) [reply addons/account_payment/views/account_portal_templates.xml:67]: If so, you should define what is `pending_txs` because it currently evaluate as `None`.

File: addons/payment/views/portal_templates.xml [hunk 7]
```diff
@@ -16,6 +16,21 @@
         </xpath>
     </template>
 
+    <template id="portal_my_home_payment" name="Show Payment Method" customize_show="True" inherit_id="portal.portal_my_home" priority="60">
```
Comment (AntoineVDV) [inline addons/payment/views/portal_templates.xml:19]: It's cool stuff to add a card for payment methods :)

I believe you can now remove the `payment.pay_meth_link` template since it does the same job but looks worse.

File: addons/payment/views/portal_templates.xml [hunk 8]
```diff
@@ -16,6 +16,21 @@
...
+        </xpath>
+        <div id="portal_client_category" position="inside">
+            <t t-call="portal.portal_docs_entry">
+                <t t-set="icon" t-value="'/payment/static/img/payment-methods.svg'"/>
+                <t t-set="title">Payment method</t>
+                <t t-set="text">Add, remove or modify your payment methods</t>
+                <t t-set="url" t-value="'/my/payment_method'"/>
+                <t t-set="config_card" t-value="True"/>
+            </t>
+        </div>
```
Comment (AntoineVDV) [inline addons/payment/views/portal_templates.xml:31]: This should be shown under the same conditions as the `payment.pay_meth_link` template. If no token exists and no provider is enabled, it makes no sense to show this card.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Hello, please find my review comments for the Payment and eCommerce scopes below :)

File: addons/account_payment/views/account_portal_templates.xml (hunk 6 above)
Comment (chgo-odoo) [reply addons/account_payment/views/account_portal_templates.xml]: Good catch. The issue was more the fact we called the same element twice but with a different styling

**Comment by chgo-odoo** [comment]
Hello @AntoineVDV, thanks for reviewing. I made the requested changes, could you have a look ?

Thanks

File: addons/payment/views/portal_templates.xml [hunk 9]
```diff
@@ -323,4 +308,23 @@
...
+            <t t-set="portal_client_category_enable" t-value="True"/>
+        </xpath>
+        <div id="portal_client_category" position="inside">
+            <t t-set="partner" t-value="request.env.user.partner_id"/>
+            <t t-set="providers_allowing_tokenization"
+               t-value="request.env['payment.provider'].sudo()._get_compatible_providers(request.env.company.id, partner.id, 0., force_tokenization=True, is_validation=True)"/>
+            <t t-set="existing_tokens" t-value="partner.payment_token_ids + partner.commercial_partner_id.sudo().payment_token_ids"/>
+            <t t-if="providers_allowing_tokenization or existing_tokens" t-call="portal.portal_docs_entry">
+                <t t-set="icon" t-value="'/payment/static/img/payment-methods.svg'"/>
+                <t t-set="title">Payment method</t>
```
Comment (AntoineVDV) [inline addons/payment/views/portal_templates.xml:322]:
```suggestion
                <t t-set="title">Payment methods</t>
```

File: addons/payment/views/portal_templates.xml [hunk 10]
```diff
@@ -323,4 +308,23 @@
...
+                <t t-set="text">Add, remove or modify your payment methods</t>
```
Comment (AntoineVDV) [inline addons/payment/views/portal_templates.xml:323]:
```suggestion
                <t t-set="text">Manage your payment methods</t>
```

File: addons/payment/views/portal_templates.xml [hunk 11]
```diff
@@ -323,4 +308,23 @@
         </div>
     </template>
 
+    <template id="portal_my_home_payment" name="Show Payment Method" customize_show="True" inherit_id="portal.portal_my_home" priority="60">
```
Comment (AntoineVDV) [inline addons/payment/views/portal_templates.xml:311]: `<!-- "Manage payment methods" card on /my -->`

**Review by AntoineVDV — APPROVED** [review]

File: addons/account/controllers/portal.py [hunk 12]
```diff
@@ -14,9 +14,13 @@ class PortalAccount(CustomerPortal):
     def _prepare_home_portal_values(self, counters):
         values = super()._prepare_home_portal_values(counters)
         if 'invoice_count' in counters:
-            invoice_count = request.env['account.move'].search_count(self._get_invoices_domain()) \
+            invoice_count = request.env['account.move'].search_count([('move_type', 'in', ('out_invoice', 'out_refund'))] + self._get_invoices_domain()) \
```
Comment (william-andre) [inline addons/account/controllers/portal.py:17]: Why are we not including the receipts anymore?

Comment (fdamhaut) [reply addons/account/controllers/portal.py:17]: It seems to be an oversight from my part, will fix it. 

Functionally, I don't exactly know what they rerpresent, but i guess out_receipt are linked to invoice and in_receipt to bills ?

Comment (william-andre) [reply addons/account/controllers/portal.py:17]: It's basically invoices without a mandatory partner, and you are right for the types: `out` says with `out` and `in` stays with `in`

**Review by william-andre — APPROVED** [review]
Okay for the accounting part! :+1: 
Didn't review the XML changes in detail but it looks alright at first glance, didn't see anything weird.

File: addons/sale/views/sale_portal_templates.xml [hunk 13]
```diff
@@ -19,18 +19,28 @@
...
+            <t t-set="portal_client_category_enable" t-value="True"/>
+            <t t-set="portal_alert_category_enable" t-value="True"/>
+        </xpath>
+        <div id="portal_alert_category" position="inside">
             <t t-call="portal.portal_docs_entry">
-                <t t-set="title">Quotations</t>
-                <t t-set="url" t-value="'/my/quotes'"/>
+                <t t-set="bg_color" t-value="'alert alert-primary'"/>
                 <t t-set="placeholder_count" t-value="'quotation_count'"/>
+                <t t-set="title">Quotation(s) to review</t>
```
Comment (Feyensv) [inline addons/sale/views/sale_portal_templates.xml:30]:
```suggestion
                <t t-set="title">Quotations to review</t>
```

Comment (stefanorigano) [reply addons/sale/views/sale_portal_templates.xml]: Makes sense to me, addressed here [e114976](https://github.com/odoo/odoo/pull/138077/commits/e1149768d889bd3d668884d574a51221052bce9a) 👍

**Review by Feyensv — APPROVED** [review]
Not convinced by some functional choices, but not worth fighting for.
Thanks for the little fixup :+1:

**Comment by stefanorigano** [comment]
Hello everyone 👋
Thanks for you reviews so far.

Since apparently the are no other remarks, can we  r+ this?
 
 cc. @dalaOdoo , @antonylesuisse

**Review by xavierbol — APPROVED** [review]

**Comment by chgo-odoo** [comment]
Hi @xmo-odoo 👋

I've fixed the layout of the navbar so that it doesn't break at a scale. These changes have been addressed here :
https://github.com/odoo/odoo/pull/138077/commits/79e4e638cb9ee0d9a31dde538c7063fc9e9bbc99

Could you have a look at it?

Thanks

File: addons/payment/views/portal_templates.xml [hunk 14]
```diff
@@ -2,18 +2,35 @@
...
+             Only show the link if a token can be created or if one already exists
             <div t-if="providers_allowing_tokenization or existing_tokens"
                  class='manage_payment_method mt16'>
                 <a href="/my/payment_method">Manage payment methods</a>
             </div>
         </xpath>
+        -->
+    </template>
+
+    <template id="portal_my_home_payment" name="Show Payment Method" customize_show="True" inherit_id="portal.portal_my_home" priority="60">
```
Comment (tde-banana-odoo) [inline addons/payment/views/portal_templates.xml:21]: This template is redefined below (with same name bit slightly different), I believe it is an error ?

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Review by reth-odoo — COMMENTED** [review]
Hey :wave: Seems alright overall, mostly small suggestions.

Just one big comment about domains for the new category, but let me know if I missed something

File: addons/payment/views/portal_templates.xml [hunk 15]
```diff
@@ -2,18 +2,20 @@
...
             <t t-set="existing_tokens" t-value="partner.payment_token_ids + partner.commercial_partner_id.sudo().payment_token_ids"/>
-            <!-- Only show the link if a token can be created or if one already exists -->
+             Only show the link if a token can be created or if one already exists
             <div t-if="providers_allowing_tokenization or existing_tokens"
                  class='manage_payment_method mt16'>
                 <a href="/my/payment_method">Manage payment methods</a>
             </div>
         </xpath>
+        -->
     </template>
```
[the remark this answers is not in the document]
Comment (chgo-odoo) [reply addons/payment/views/portal_templates.xml:19]: This will be removed in another PR with an upgrade script so we keep it like that for now

File: addons/sale_timesheet/views/sale_timesheet_portal_templates.xml [hunk 16]
```diff
@@ -90,17 +90,19 @@
...
-            <li class="list-group-item flex-grow-1">
-                <div class="btn-toolbar flex-sm-nowrap justify-content-center">
-                    <div class="btn-group mb-1">
-                        <t t-set="search_value" t-value="invoice.name"/>
-                        <t t-if="invoice.state == 'draft'" t-set="search_value" t-value="invoice.id"/>
-                        <a t-if="invoice.move_type == 'out_invoice' and invoice.state in ('draft', 'posted') and invoice.timesheet_count > 0"
-                        target="_blank" t-att-href="'/my/timesheets?search_in=invoice&amp;search=%s' % search_value">View Timesheets</a>
+        <xpath expr="//t[@t-set='entries']/div/div/div[hasclass('o_download_pdf')]" position="inside">
+            <t t-if="invoice.timesheet_count > 0">
+                <li class="list-group-item flex-grow-1 border-0">
```
[the remark this answers is not in the document]
Comment (chgo-odoo) [reply addons/sale_timesheet/views/sale_timesheet_portal_templates.xml:95]: Doing so would break the layout since `list-group-item` adds a padding

File: addons/website_crm_partner_assign/views/website_crm_partner_assign_templates.xml [hunk 17]
```diff
@@ -663,42 +662,36 @@
...
-                                    </div>
-                                    <div t-if="opportunity.mobile">
-                                        <span class="fa fa-mobile fa-fw" role="img" aria-label="Mobile" title="Mobile"/>
-                                        <span itemprop="mobile" t-field="opportunity.mobile"/>
-                                    </div>
-                                    <div t-if="opportunity.email_from">
-                                        <span class="fa fa-envelope fa-fw" role="img" aria-label="Email" title="Email"/>
-                                        <span itemprop="email" t-field="opportunity.email_from"/>
+                                    <div class="d-flex justify-content-start align-items-center gap-2 mb-2" t-if="opportunity.partner_name or opportunity.contact_name">
+                                        <img t-if="opportunity.partner_id.avatar_1024" class="o_avatar o_portal_contact_img rounded" t-att-src="image_data_uri(opportunity.partner_id.avatar_1024)"
```
[the remark this answers is not in the document]
Comment (chgo-odoo) [reply addons/website_crm_partner_assign/views/website_crm_partner_assign_templates.xml]: Since there is a `t-if`, the image won't be displayed if no partner exists, unless I'm getting something wrong?

[the remark this answers is not in the document]
Comment (thiryr) [reply addons/website_crm_partner_assign/views/website_crm_partner_assign_templates.xml]: Oh right, it will be falsy even if there's no partner. Indeed this is fine.

File: addons/project/views/project_portal_project_task_templates.xml [hunk 18]
```diff
@@ -189,19 +189,19 @@
...
+                                                <h6 class="mb-0" t-field="user.name"></h6>
                                             </div>
                                             <div t-out="user" t-options='{"widget": "contact", "fields": ["email", "phone"]}'/>
                                         </div>
                                     </t>
                                 </div>
                                 <div class="col-12 d-flex flex-column" t-if="task.partner_id">
                                     <h6><small class="text-muted">Customer</small></h6>
-                                    <t t-foreach="task.partner_id" t-as="partner">
+                                    <t t-id="task.partner_id">
```
Comment (fdamhaut) [inline addons/project/views/project_portal_project_task_templates.xml:200]: Oops this is meant to be a t-if

File: addons/account/controllers/portal.py [hunk 19]
```diff
@@ -16,9 +16,13 @@ class PortalAccount(CustomerPortal):
     def _prepare_home_portal_values(self, counters):
         values = super()._prepare_home_portal_values(counters)
         if 'invoice_count' in counters:
-            invoice_count = request.env['account.move'].search_count(self._get_invoices_domain()) \
+            invoice_count = request.env['account.move'].search_count(self._get_invoices_domain('in'), limit=1) \
```
Comment (thiryr) [inline addons/account/controllers/portal.py:19]: Wait if we're doing limit=1 won't it always be 0 or 1?

Comment (fdamhaut) [reply addons/account/controllers/portal.py:19]: > Wait if we're doing limit=1 won't it always be 0 or 1?
Cards were changed to not show number anymore, (not for all but most of them) if there is at least 1 invoice/bill the card will be shown so there is no point in knowing there is more.

Limit could be added everywhere but for the sake of (finally) merging this PR I only changed where I made changes.

Comment (thiryr) [reply addons/account/controllers/portal.py:19]: Makes sense, thanks

**Review by reth-odoo — APPROVED** [review]
Seems ok thanks for the changes

@tde-banana-odoo freezodoo r+ ?

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge
